### ISSUE TITLE
Add a New Trait: `MarshalableVariant`

### ISSUE THREAD

**Issue by trent-reed** [tpm-rs/tpm-rs#84]
# Summary

The `Marshalable` crate has some private functions that leak into being accessible to other types in the case of `enum` items currently. I propose that we officially say that these functions can be accessed from the public API, and say they are under a new trait named `MarshalableEnum`.

# Problem Statement

Today, we have the following trait:

```rs
pub trait Marshalable {
    fn try_unmarshal(buffer: &mut UnmarshalBuf) -> TpmRcResult<Self>
    where
        Self: Sized;
    fn try_marshal(&self, buffer: &mut [u8]) -> TpmRcResult<usize>;
}
```

There's actually a few issues here (for example, `TpmRcResult` is not the right error type; it's actually supposed to be a different error type for client vs server - we can solve this with an indirect error type; I'll open a separate issue for that), but for now let's just focus on what this issue is talking about - the accidentally visible enum functions.

When implementing `Marshalable` for an enum, the proc-macro define the following private functions on the type:

```rs
impl $Type {
    fn discriminant(&self) -> #Primitive;
    fn try_marshal_variant(&self, buffer: &mut [u8]) -> TpmRcResult<usize>;
    fn try_unmarshal_variant(selector: #Primitive, buffer: &mut UnmarshalBuf) -> TpmRcResult<Self>;
}
```

Now this is okay so far, it makes sense to separate the functions out, no harm no foul.

Where things become questionable is when you have another separate type that uses the enum, and does ***NOT*** have the discriminant placed right before the enum data.

Like this:

```rs
pub struct TpmsAttest {
    pub magic: TPM2Generated,
    // NOTE: There is a hidden discriminant here!
    pub qualified_signer: Tpm2bName,
    pub extra_data: Tpm2bData,
    pub clock_info: TpmsClockInfo,
    pub firmware_version: u64,
    pub attested: TpmuAttest, // This is the enum for the discriminant!
}
```

This creates a few problems:

1. It currently makes it impossible for us to mark this up and use the `Marshalable` proc-macro in a sensible way.
2. Since these functions are private, it's impossible to define an enum in one module, and a structure with an out-of-line discriminant for that enum in a different module (because the functions are private).
3. By extension, if a vendor wishes to provide custom types/requests/responses/commands (something that I understand is actually quite common in TPMs), then this format currently does not work cross-crate.

This is a path that I believe our `Marshalable` proc-macro should support. I want to highlight here is that calling functions like `discriminant()`, `try_marshal_variant()`, and `try_unmarshal_variant()` must be possible from outside of the private scope of the marshalable type, and in fact from outside of the current crate even (for vendor-specific types, that we don't want defined in `tpm2-rs-base`).

# Solution: `MarshalableEnum` Trait

What I'd like to do here is make these functions an official part of our API, and that can be accomplished with a new trait:

```rs
// Again, for now we are ignoring the incorrect error type - one thing at a time.
pub trait MarshalableEnum: Marshalable {
  type Selector: Sized;
  fn discriminant(&self) -> Self::Selector;
  fn try_unmarshal_variant(selector: Self::Selector, buffer: &mut UnmarshalBuf) -> TpmRcResult<Self>;
  fn try_marshal_variant(&self, buffer: &mut [u8]) -> TpmRcResult<usize>;
}
```

For the time being, I'd really like to do this so that I can unblock splitting up `tpm2-rs-base` - so I'm going to suggest something a little hacky for the time being, but also a path towards it being not hacky at all in the future:

1. First, let's just have `Marshalable` proc-macro define this new `MarshalableEnum` trait as well as the original `Marshalable` trait to unblock breaking up `tpm2-rs-base`. (This is the hacky thing, this is actually not ideal because it hides that enums implement this trait.)
2. Later, we will actually properly have a separate proc-macro (also provided by `tpm2-rs-marshalable`, since it's closely related I see no point in creating new crates here). This way we can eventually derive enums like `#[derive(Marshalable, MarshalableEnum)]` to do the more common rusty thing of making the derives visible from the attributes.
4. Finally, we should support some way of marking up these out-of-line discriminant types so that it's more obvious where the discriminant is being written to, as well as no longer needing custom `Marshalable` implementations on the types

# Example Eventual Final State:

```rs
#[derive(Marshalable, MarshalableEnum)]
pub enum TpmuAttest {
    ...
}

#[derive(Marshalable)]
pub struct TpmsAttest {
    pub magic: TPM2Generated,
    
    // Private because it's only used for marshaling/unmarshaling.
    // But this makes it really clear that this is just a write/read location.
    // Note: I'm actually not sold on it looking exactly like this,
    //       I'm just providing an example for a potential future.
    _discriminant: DiscriminantMarker<TpmuAttest>,

    pub qualified_signer: Tpm2bName,
    pub extra_data: Tpm2bData,
    pub clock_info: TpmsClockInfo,
    pub firmware_version: u64,
    #[marshal(discriminant=_discriminant)]
    pub attested: TpmuAttest,
}
```

EDIT: Added `DiscriminantMarker` to the example above, imagine this as a zero-sized type that is only used for struct markup. Kinda like `PhantomData`.

**Comment by oddcoder** [comment]
So I spent some time thinking about this problem, and here are my thoughts:

I understand that you claim that the with separate discriminate and enum may not work with the current implementation. And I understand that you claim that exposing `MarshalableEnum` will help handling that case.

But I was thinking, do we really need to expose all that API? 

Shouldn't this be enough?

```rust
pub trait MarshalableEnum: Marshalable {
  type Selector: Sized;
  fn discriminant(&self) -> Self::Selector;
  }
```

**Comment by oddcoder** [comment]
Also I had another critique but it is not related to this issue, only related to the example, I only put it here so maybe we can let the idea sink in for longer, and give it more iterations until it is time to resolve it.

```rust
#[derive(Marshalable)]
pub struct TpmsAttest {
    pub magic: TPM2Generated,
    _discriminant: DiscriminantMarker<TpmuAttest>,
    pub qualified_signer: Tpm2bName,
    pub extra_data: Tpm2bData,
    pub clock_info: TpmsClockInfo,
    pub firmware_version: u64,
    #[marshal(discriminant=_discriminant)]
    pub attested: TpmuAttest,
}
```

That makes it harder to initialize  `TpmsAttest` by `let` binding without a `new` method.
Objectively ` _discriminant: DiscriminantMarker<TpmuAttest>` offers no type information for once (unless I misunderstand its role) and it offers no data storage since it is zst. The only information it offers is location information, in particular that it is before `qualified_signer` and after `magic`.  We can encode the same information without needing this marker.

I think of two approaches:

```rust
#[derive(Marshalable)]
pub struct TpmsAttest {
    pub magic: TPM2Generated,
    #[marshal(marker = _discriminant)]
    pub qualified_signer: Tpm2bName,
    pub extra_data: Tpm2bData,
    pub clock_info: TpmsClockInfo,
    pub firmware_version: u64,
    #[marshal(discriminant=_discriminant)]
    pub attested: TpmuAttest,
}
```
This is the least verbose, and the most natural from user perspective,  but there couple of edge cases and will require custom representation of `syn` structure because in rust they do not have free floating attributes.


The other approach will be:


```rust
#[derive(Marshalable)]
#[marshal(marker(after = qualified_signer, name =magic))]
pub struct TpmsAttest {
    pub magic: TPM2Generated,
    pub qualified_signer: Tpm2bName,
    pub extra_data: Tpm2bData,
    pub clock_info: TpmsClockInfo,
    pub firmware_version: u64,
    #[marshal(discriminant=_discriminant)]
    pub attested: TpmuAttest,
}
```

This is more verbose and a bit unnatural, but it is easy to implement. The only advantage both approaches share is that `TpmsAttest` can be constructed without any kind of constructor method.

**Comment by trent-reed** [comment]
> Shouldn't this be enough?
> 
> ```rs
> pub trait MarshalableEnum: Marshalable {
>   type Selector: Sized;
>   fn discriminant(&self) -> Self::Selector;
>   }
> ```

I don't see how. Today, we need to manually implement the marshalling functions to get this to work, and they look like this:

```rs
impl Marshalable for TpmsAttest {
    fn try_marshal(&self, buffer: &mut [u8]) -> TpmRcResult<usize> {
        let mut written = 0;
        written += self.magic.try_marshal(&mut buffer[written..])?;
        written += self
            .attested
            .discriminant()
            .try_marshal(&mut buffer[written..])?;
        written += self.qualified_signer.try_marshal(&mut buffer[written..])?;
        written += self.extra_data.try_marshal(&mut buffer[written..])?;
        written += self.clock_info.try_marshal(&mut buffer[written..])?;
        written += self.firmware_version.try_marshal(&mut buffer[written..])?;
        written += self.attested.try_marshal_variant(&mut buffer[written..])?;
        Ok(written)
    }

    fn try_unmarshal(buffer: &mut UnmarshalBuf) -> TpmRcResult<Self> {
        let magic = TPM2Generated::try_unmarshal(buffer)?;
        let selector = u16::try_unmarshal(buffer)?;
        Ok(TpmsAttest {
            magic,
            qualified_signer: Tpm2bName::try_unmarshal(buffer)?,
            extra_data: Tpm2bData::try_unmarshal(buffer)?,
            clock_info: TpmsClockInfo::try_unmarshal(buffer)?,
            firmware_version: u64::try_unmarshal(buffer)?,
            attested: TpmuAttest::try_unmarshal_variant(selector, buffer)?,
        })
    }
}
```

You can see how we have to call `try_marshal_variant` and `try_unmarshal_variant`:

* `try_marshal_variant` will only marshal the variant data.
* `try_unmarshal_variant` needs to be told what the discriminant is so that it can interpret the bytes in the correct way.

The current `try_marshal` and `try_unmarshal` for enums basically is just:

1. Assume the discriminant comes first.
2. Assume the remainder is any data for the variant.

We could disallow this, and say that the only thing these types may marshal/unmarshal is their variant data, and maybe that works for `try_marshal`, but it doesn't work for `try_unmarshal` since the function parameters are different (we have to be able to tell it what the format of the bytes it's unmarshalling is).

I do also see a future where we don't implement `Marshalable` derives on enums, and they always have an explicit location for their disriminant in the parent structure. In that case there's no need for `Marshalable`, and we only need `MarshalableEnum`, which feels a tad cleaner to me since we make the discriminant location clearly stated in the type. (But I'm not certain this is wise, or that it won't cause code cleanliness issues, I just want to say I think it's possible.)

And re-using `Marshalable` sounds like it wont work - since we need different parameters for unmarshalling. I agree with the existence of `try_marshal_variant` and `try_unmarshal_variant`, and I like to have names different from the original marshalable since these functions work in a slightly different way (skipping the discriminant data).

We could maybe do it something like this (disclaimer: I don't like this):

```
trait MarshalableEnum {
    type Selector: Sized;
}

trait Marshalable {
  fn try_marshal(&self, buffer: &mut [u8]) -> TpmRcResult<usize>;
  fn try_unmarshal(buffer: &mut UnmarshalBuf) -> TpmRcResult<Self>;
  fn try_marshal_variant(&self, buffer: &mut [u8]) -> TpmRcResult<usize>
      where Self: MarshalableEnum;
  fn try_unmarshal(selector: Self::Selector, buffer: &mut UnmarshalBuf) -> TpmRcResult<Self>
      where Self: MarshalableEnum;
}
```

But I can't find a way out of needing these extra functions, and if we're going to do that, why not just put them in the sensible place (`MarshalableEnum`) and treat them like we would any other traits?

**Comment by trent-reed** [comment]
> Also I had another critique but it is not related to this issue, only related to the example

Yeah, I just wanted to show it has a possible future towards removing the requirement of manually having to implement the `Marshalable` trait for types that do this. I think that should be a requirement of whatever path we go down - I'm open to discussing different options, but let's save that for another issue.

**Comment by oddcoder** [comment]
> But I can't find a way out of needing these extra functions

I would say this, i don't particularly like it, but I cannot find any other way to do it unfortunately. So I will not get in your way of doing it. I don't think it makes sense to prolong the discussion if we cannot find a better way. Feel free to bundle them into enum. 

> EDIT: I wanted to add that I kinda like the idea of using attributes for the markers; I was originally thinking of it like a phantom data situation, but then we do have to make it pub since you have to assign the marker, and it's kind of annoying to do since it's an implementation detail, so I agree with your points there. Anyways, we'll put all the ways together in an issue and see what works best - but I'm liking the attribute over the type one past the discriminant the best so far, I think.

Sure if you like feel free to do it after, If not I can do, I prefer the version with ` #[marshal(marker = _discriminant)]` but it needs different `struct` representation (I think) since this attribute could be  the last thing in the struct, and curly braces cannot have attribute.

**Comment by oddcoder** [comment]
one more thing, would it make sense if we later some how move those discussions to docs as documents? I think they would offer lots of information for newcomers.

**Comment by trent-reed** [comment]
> Sure if you like feel free to do it after, If not I can do, I prefer the version with ` #[marshal(marker = _discriminant)]` but it needs different `struct` representation (I think) since this attribute could be the last thing in the struct, and curly braces cannot have attribute.

It actually can't be, because the discriminant for an enum is defined to always come before the data for the enum, so it may never be the last thing in a struct unless it's a standalone simple constant enum type with no variant data. So I don't think this can be wrong ever, but I probably need to think of it more.

From TPM2 Spec Part2 (4.9.3: Union Instance): 

> The selector may be any of the parameters that
occur in the structure before the union instance.

So it may never be the last element - makes sense, otherwise you can't find the selector so you don't know how to interpret the bytes before it.

**Comment by trent-reed** [comment]
> I would say this, i don't particularly like it, but I cannot find any other way to do it unfortunately. So I will not get in your way of doing it. I don't think it makes sense to prolong the discussion if we cannot find a better way. Feel free to bundle them into trait.

I think we can re-evaluate this and see if we have better ideas before we do our first crate publish. We need to re-evaluate our crate layout anyways. So for now I'll just proceed with the plan outlined by this document (and #79 can remove the explicit `discriminant` function and `Selector` type).

I did have a thought though, it does seem like maybe this trait should be named `MarshalableVariant` so that the trait name matches a little better with the functions it provides - just thinking out loud here. I'll produce a PR and we'll hash it out in there.

**Comment by oddcoder** [comment]
solved by #89

**Comment by trent-reed** [comment]
I didn't realize this was closed.

This is only partially solved, actually. The main issue of the proc-macro not supporting selectors separated from the data is still present.

I'd like to open this until I can file a different bug that deals with that so we don't lose the context. So re-opening until I have time to do that.

**Comment by oddcoder** [comment]
+1 for different issue.

**Comment by trent-reed** [comment]
Closed for #106 and #107
